File: tests/test_indefinite_ber.py
"""Tests for indefinite length BER encoding/decoding roundtrip."""
from TestModule import Person, Department, Company, Config, Certificate, Status


def test_person_ber_indefinite_roundtrip():
    p = Person(name="Alice", age=30, active=True, status=0)
    encoded = p.encode_ber_indefinite()
    assert encoded[:2] == bytes([0x30, 0x80])
    assert encoded[-2:] == bytes([0x00, 0x00])
    decoded = Person.decode_ber_indefinite(encoded)
    assert decoded.name == "Alice"
    assert decoded.age == 30
    assert decoded.active is True
    assert decoded.status == 0


def test_department_ber_indefinite_roundtrip():
    d = Department(name="Engineering", people=[])
    encoded = d.encode_ber_indefinite()
    assert encoded[:2] == bytes([0x30, 0x80])
    assert encoded[-2:] == bytes([0x00, 0x00])
    decoded = Department.decode_ber_indefinite(encoded)
    assert decoded.name == "Engineering"


def test_company_ber_indefinite_roundtrip():
    c = Company(name="Acme Corp", departments=[])
    encoded = c.encode_ber_indefinite()
    assert encoded[:2] == bytes([0x30, 0x80])
    assert encoded[-2:] == bytes([0x00, 0x00])
    decoded = Company.decode_ber_indefinite(encoded)
    assert decoded.name == "Acme Corp"


def test_config_ber_indefinite_defaults():
    c = Config()
    encoded = c.encode_ber_indefinite()
    assert encoded[:2] == bytes([0x30, 0x80])
    assert encoded[-2:] == bytes([0x00, 0x00])
    decoded = Config.decode_ber_indefinite(encoded)
    assert decoded.enabled is True
    assert decoded.count == 42
    assert decoded.label == "default"


def test_config_ber_indefinite_with_values():
    c = Config(enabled=False, count=100, label="custom")
    encoded = c.encode_ber_indefinite()
    assert encoded[:2] == bytes([0x30, 0x80])
    assert encoded[-2:] == bytes([0x00, 0x00])
    decoded = Config.decode_ber_indefinite(encoded)
    assert decoded.enabled is False
    assert decoded.count == 100
    assert decoded.label == "custom"


def test_ber_indefinite_inner_matches_der_inner():
    """Verify indefinite inner content matches DER inner content."""
    p = Person(name="Bob", age=25, active=False, status=1)
    der = p.encode_der()
    indefinite = p.encode_ber_indefinite()
    der_inner = der[2:]
    indefinite_inner = indefinite[2:-2]
    assert indefinite_inner == der_inner


def test_ber_indefinite_decode_matches_der_decode():
    """Verify decoding indefinite produces same result as DER decode."""
    p = Person(name="Charlie", age=40, active=True, status=0)
    der = p.encode_der()
    indefinite = p.encode_ber_indefinite()
    decoded_der = Person.decode_der(der)
    decoded_indefinite = Person.decode_ber_indefinite(indefinite)
    assert decoded_der.name == decoded_indefinite.name
    assert decoded_der.age == decoded_indefinite.age
    assert decoded_der.active == decoded_indefinite.active
    assert decoded_der.status == decoded_indefinite.status


def test_ber_indefinite_nested():
    """Test indefinite encoding with SEQUENCE/SET types."""
    c = Company(name="Nested Corp", departments=[])
    encoded = c.encode_ber_indefinite()
    assert encoded[:2] == bytes([0x30, 0x80])
    assert encoded[-2:] == bytes([0x00, 0x00])
    decoded = Company.decode_ber_indefinite(encoded)
    assert decoded.name == "Nested Corp"
    assert len(decoded.departments) == 0


def test_ber_indefinite_department_with_people():
    """Test indefinite encoding with list of primitive types."""
    p = Person(name="Alice", age=30, active=True, status=0)
    encoded = p.encode_ber_indefinite()
    assert encoded[:2] == bytes([0x30, 0x80])
    assert encoded[-2:] == bytes([0x00, 0x00])
    decoded = Person.decode_ber_indefinite(encoded)
    assert decoded.name == "Alice"
    assert decoded.age == 30
    assert decoded.active is True
    assert decoded.status == 0

File: tests/vectors/indefinite_ber/test_indefinite_ber.rs
//! ```cargo
//! [dependencies]
//! asnvil-runtime-rust = { path = "__REPO_ROOT__/asnvil-runtime-rust" }
//! num-bigint = "0.4"
//! ```

#[path = "./TestModule.rs"]
mod generated;

use generated::{Person, Department, Company, Config};
use num_bigint::BigInt;

fn main() {
    test_person_ber_indefinite_roundtrip();
    test_department_ber_indefinite_roundtrip();
    test_company_ber_indefinite_roundtrip();
    test_config_ber_indefinite_defaults();
    test_config_ber_indefinite_with_values();
    test_ber_indefinite_inner_matches_der_inner();
    test_ber_indefinite_decode_matches_der_decode();
    test_ber_indefinite_nested();
    test_ber_indefinite_department_with_people();
    println!("\nAll Indefinite BER integration tests passed!");
}

fn test_person_ber_indefinite_roundtrip() {
    let person = Person {
        name: "Alice".to_string(),
        age: BigInt::from(30),
    };
    let encoded = person.encode_ber_indefinite().expect("Failed to encode");
    let decoded = Person::decode_ber_indefinite(&encoded).expect("Failed to decode");
    assert_eq!(decoded.name, "Alice");
    assert_eq!(decoded.age, BigInt::from(30));
    println!("PASS: test_person_ber_indefinite_roundtrip");
}

fn test_department_ber_indefinite_roundtrip() {
    let dept = Department {
        name: "Engineering".to_string(),
        id: BigInt::from(101),
    };
    let encoded = dept.encode_ber_indefinite().expect("Failed to encode");
    let decoded = Department::decode_ber_indefinite(&encoded).expect("Failed to decode");
    assert_eq!(decoded.name, "Engineering");
    assert_eq!(decoded.id, BigInt::from(101));
    println!("PASS: test_department_ber_indefinite_roundtrip");
}

fn test_company_ber_indefinite_roundtrip() {
    let dept = Department {
        name: "Engineering".to_string(),
        id: BigInt::from(101),
    };
    let company = Company {
        name: "TechCorp".to_string(),
        departments: vec![dept],
    };
    let encoded = company.encode_ber_indefinite().expect("Failed to encode");
    let decoded = Company::decode_ber_indefinite(&encoded).expect("Failed to decode");
    assert_eq!(decoded.name, "TechCorp");
    assert_eq!(decoded.departments.len(), 1);
    println!("PASS: test_company_ber_indefinite_roundtrip");
}

fn test_config_ber_indefinite_defaults() {
    let config = Config {
        timeout: None,
        retries: None,
    };
    let encoded = config.encode_ber_indefinite().expect("Failed to encode");
    let decoded = Config::decode_ber_indefinite(&encoded).expect("Failed to decode");
    assert!(decoded.timeout.is_none());
    assert!(decoded.retries.is_none());
    println!("PASS: test_config_ber_indefinite_defaults");
}

fn test_config_ber_indefinite_with_values() {
    let config = Config {
        timeout: Some(BigInt::from(30)),
        retries: Some(BigInt::from(3)),
    };
    let encoded = config.encode_ber_indefinite().expect("Failed to encode");
    let decoded = Config::decode_ber_indefinite(&encoded).expect("Failed to decode");
    assert_eq!(decoded.timeout, Some(BigInt::from(30)));
    assert_eq!(decoded.retries, Some(BigInt::from(3)));
    println!("PASS: test_config_ber_indefinite_with_values");
}

fn test_ber_indefinite_inner_matches_der_inner() {
    let person = Person {
        name: "Bob".to_string(),
        age: BigInt::from(25),
    };
    let ber_encoded = person.encode_ber_indefinite().expect("Failed to encode BER");
    let der_encoded = person.encode_der().expect("Failed to encode DER");
    // The inner content should be similar, though BER has indefinite length markers
    assert!(ber_encoded.len() > der_encoded.len());
    println!("PASS: test_ber_indefinite_inner_matches_der_inner");
}

fn test_ber_indefinite_decode_matches_der_decode() {
    let person = Person {
        name: "Charlie".to_string(),
        age: BigInt::from(40),
    };
    let ber_encoded = person.encode_ber_indefinite().expect("Failed to encode BER");
    let der_encoded = person.encode_der().expect("Failed to encode DER");
    let ber_decoded = Person::decode_ber_indefinite(&ber_encoded).expect("Failed to decode BER");
    let der_decoded = Person::decode_der(&der_encoded).expect("Failed to decode DER");
    assert_eq!(ber_decoded, der_decoded);
    println!("PASS: test_ber_indefinite_decode_matches_der_decode");
}

fn test_ber_indefinite_nested() {
    let dept = Department {
        name: "HR".to_string(),
        id: BigInt::from(200),
    };
    let company = Company {
        name: "MegaCorp".to_string(),
        departments: vec![dept],
    };
    let encoded = company.encode_ber_indefinite().expect("Failed to encode");
    let decoded = Company::decode_ber_indefinite(&encoded).expect("Failed to decode");
    assert_eq!(decoded.name, "MegaCorp");
    assert_eq!(decoded.departments[0].name, "HR");
    println!("PASS: test_ber_indefinite_nested");
}

fn test_ber_indefinite_department_with_people() {
    let dept = Department {
        name: "Dev".to_string(),
        id: BigInt::from(300),
    };
    let encoded = dept.encode_ber_indefinite().expect("Failed to encode");
    let decoded = Department::decode_ber_indefinite(&encoded).expect("Failed to decode");
    assert_eq!(decoded.name, "Dev");
    assert_eq!(decoded.id, BigInt::from(300));
    println!("PASS: test_ber_indefinite_department_with_people");
}